### PR TITLE
Session-persistent cache

### LINKED ISSUE #1221: Task output does not seem to be cached (despite `cache_for` argument being present)
_(I'm using prefect version `0.5.5`.)_

Here is a simple flow containing a single task that has the `cache_for` argument set.

I would expect this flow to take ~2s the first time it runs and then pretty much complete instantaneously on subsequent runs.
```python
from prefect import task, Flow
from datetime import timedelta
import time

@task(cache_for=timedelta(hours=1))
def return_the_answer():
    time.sleep(2)
    return 42

with Flow("answer") as flow:
    return_the_answer()

flow.run()
```

Here is the output from running in a Jupyter notebook using `%%time`, which confirms that the flow takes ~2s to complete.
```
[2019-07-10 20:45:09,227] INFO - prefect.FlowRunner | Beginning Flow run for 'answer'
[2019-07-10 20:45:09,229] INFO - prefect.FlowRunner | Starting flow run.
[2019-07-10 20:45:09,235] INFO - prefect.TaskRunner | Task 'return_the_answer': Starting task run...
[2019-07-10 20:45:11,240] INFO - prefect.TaskRunner | Task 'return_the_answer': finished task run for task with final state: 'Cached'
[2019-07-10 20:45:11,241] INFO - prefect.FlowRunner | Flow run SUCCESS: all reference tasks succeeded
CPU times: user 14.6 ms, sys: 4.24 ms, total: 18.8 ms
Wall time: 2.02 s
<Success: "All reference tasks succeeded.">
```

The strange thing is that subsequent runs take exactly the same amount of time but I'd expect the result to be cached so they should complete much more quickly.

It's quite possible that this is a misunderstanding on my part rather than a bug. Am I supplying the `cache_for` argument in the wrong place? (I couldn't find an example in the docs which applies it to tasks defined via the `@task` decorator.) Any clarifications would be appreciated. Many thanks!

### PR DESCRIPTION
**Pull request by cicdw**
**Thanks for contributing to Prefect!**

Please describe your work and make sure your PR:

- [x] adds new tests (if appropriate)
- [x] updates `CHANGELOG.md` (if appropriate)
- [x] updates docstrings for any new functions or function arguments, including `docs/outline.toml` for API reference docs (if appropriate)

Note that your PR will not be reviewed unless all three boxes are checked.

## What does this PR change?
This PR implements a cache which _persists_ for the duration of a Python session (when used in the absence of Cloud).  Additionally, allows for the use of `cache_key`s when using the open source engine alone.  Note that this _also_ allows users to "seed" the cache by simply updating `prefect.context.caches` with an appropriate dictionary!

## Why is this PR important?
Closes #1221

### CODE REVIEW

**Comment by maxalbert** [comment]
While playing with this it just occurred to me that sadly this solution won't work in its current form.

The reason is that the cache looks at the function name (or a user-provided key) but does not take the implementation of the function into account. Thus when the user tweaks the source code in an interactive python session, prefect will still return the previously cached result, which may now be wrong.

Off the top of my head I can think of two quick-fix approaches to solve this.

- Internally append a counter to the task's `cache_id` whenever the task-to-be-cached is re-defined in the interactive session.

- Use `inspect.getsource()` to retrieve the source code of the task definition and internally append a hashed version of it to the `cache_key`. (Sadly, I don't think `inspect.getsource()` works in IPython notebooks but there is a workaround - let me know if you'd like me to dig this up.)

It's quite possible that neither of these ideas may work or be easy to implement but just thought I'd put them out there.

**Comment by maxalbert** [comment]
Here is a code snippet to reproduce this issue (with the branch for this PR):
```python
from prefect import task, Flow
from datetime import timedelta
import time

@task(cache_for=timedelta(hours=1))
def return_the_answer():
    return 42

@task
def print_the_answer(answer):
    print(f"The answer is: {answer}")

with Flow("answer") as flow:
    answer = return_the_answer()
    print_the_answer(answer)
```

This gives the expected result:
```
>>> flow.run()
[...]
The answer is: 42
```

However, if we redefine the first task to return a different result...
```python
@task(cache_for=timedelta(hours=1))
def return_the_answer():
    return 23

with Flow("answer") as flow:
    answer = return_the_answer()
    print_the_answer(answer)
```
... we get the wrong output due to the first result being cached:
```
>>> flow.run()
[...]
The answer is: 42
```

**Comment by jlowin** [comment]
@maxalbert I think that you're actually illustrating the feature working as expected; it's a *session* cache, not a *flow* cache, and you provided two cached tasks in the same session with the same `cache_key`. 

One of the reasons it was built this way is that the most common feature request around caching (other than making it easier to run locally) is to allow cached states to be used across multiple flows.

I would be very wary of building any form of implicit cache invalidation as you describe. However, there are a number of explicit ways you can inform Prefect how it should act:
- start a new session - as a matter of best practice, if you have code you know to be "wrong", you always run the risk of polluting your python session (whether using Prefect or not). When you're finished testing, run your flow in a clean session. This will be true of any code that affects persistent state.
- construct your second flow's tasks with different cache keys, so Prefect knows they don't share a cache - `Task(cache_key='my_new_key')`
- invalidate Prefect's caches - I believe `prefect.context.caches.clear()` should work
- explicitly pass an expired `Cached()` state for the cached task when running it a second time, so it won't attempt to load the global cache -- (note @cicdw this doesn't actually work for a reason I'll bring up in PR review, and may in fact not be desirable)

File: tests/core/test_flow.py [hunk 1]
```diff
@@ -47,6 +47,11 @@ def add_flow():
     return f
 
 
+@pytest.fixture
+def clear_context_cache():
```
Comment (jlowin) [inline tests/core/test_flow.py]: This might be a useful enough function to elevate to a utility file?

File: src/prefect/engine/task_runner.py [hunk 2]
```diff
@@ -574,7 +574,9 @@ def check_task_is_cached(self, state: State, inputs: Dict[str, Result]) -> State
             - ENDRUN: if the task is not ready to run
         """
         if self.task.cache_for is not None:
-            candidate_states = prefect.context.caches.get(self.task.name, [])
+            candidate_states = prefect.context.caches.get(
```
Comment (jlowin) [inline src/prefect/engine/task_runner.py]: Something about the order of operations here feels slightly off to me. I think it's that the `context.caches` check is happening before the passed `state` is checked. By the time we reach this pipeline step, we already know the task is `Pending` and ready to run -- but it seems to me if it were explicitly provided with a `Cached` state, that should take precedence over the context cache. Too complex?

Comment (cicdw) [reply src/prefect/engine/task_runner.py]: That makes sense; will update.

Comment (cicdw) [reply src/prefect/engine/task_runner.py]: Updated w/ test

**Pull request merged**